### DISCUSSION TITLE
Help me with postgresql!

### DISCUSSION THREAD

**Discussion by bookvik** [mikub/titanoboa#34] [Q&A] [answered]
```
 [{:type java.lang.IllegalArgumentException
titanoboa_1    |    :message No method in multimethod 'list-jobs' for dispatch value: org.postgresql.Driver
titanoboa_1    |    :at [clojure.lang.MultiFn getFn MultiFn.java 156]}]
```

I'm desperate. I do not understand how to start it and where to insert what

my dependencies file 

```
{:coordinates [[io.titanoboa.tasklet/http-client "0.1.1"]
               [org.postgresql/postgresql "9.4.1208"]],
 :require [[io.titanoboa.tasklet.httpclient] [titanoboa.system.jdbc]],
 :repositories {"clojars" "https://clojars.org/repo",
                "central" "https://repo1.maven.org/maven2/"},
 :import [[org.postgresql.Driver]]}
```

**Comment by mikub** [comment, marked as answer]
Hi @bookvik !
thanks for reaching out, I might look into documenting this a bit better to make sure everything is in one place and people do not get confused. Here are some examples that usually work for me to connect to a database and to query it:

Here's a sample dependencies file for postgres that should work:
```
{:coordinates [[org.postgresql/postgresql "9.4.1208"]],
 :require [[titanoboa.tasklet.jdbc]
           [titanoboa.system.jdbc]
           [titanoboa.database.postgres]],
 :repositories {"clojars" "https://clojars.org/repo",
                "central" "https://repo1.maven.org/maven2/"},
 :import [[org.postgresql.Driver]]}
```

Be also mindful of #11 (which should be fixed now though); If the driver cannot be found on classpath for some reason there is always an easy workaround - just put the  driver jar into the /lib folder.

Then you define a system that initiates a connection pool to that DB upon t-boa startup as described here:
https://github.com/mikub/titanoboa/wiki/Server-Configuration#non-core-systems

Then you can reference that connection pool from you workflow steps or you can straight away use pre-defined jdbc tasklet:
https://github.com/mikub/titanoboa-tasklets#jdbc-client-

Also if you want to write some custom code and access given data source object directly, you could do it like this from clojure code:
```clojure
(get-in @titanoboa.system/systems-state data-source-ks)
```
where `data-source-ks` is a vector containing key set that points to a system's data source, e.g. for a system called `:test-db` it would be `[:test-db :system :pool]`

Hope this helps, let me know if it does not.

**Reply by bookvik** [reply]
Thank you. it is very difficult to guess which dependencies should be registered. there are no examples and therefore have to climb the code of the engine itself.

but I see in your file several of those dependencies that I do not have, so today I will try and hope for success.

Also, the question is whether it is possible to use the latest version of the driver,  version 42 ?

**Reply by mikub** [reply]
Yeah the documentation in that regard could definitely be better, I will add a separate wiki with some examples that do work :) And also clearly separate setup for archival db vs the setup for other database connections that might be used by the workflows.

Also the version of the driver should ideally match the version of the DB, so it will depend what version of pg you are running, but it is always good  to start with the latest one...

**Reply by bookvik** [reply]
![image](https://user-images.githubusercontent.com/204606/113279012-4656de00-92eb-11eb-988b-88d48211cc79.png)

it works!!! many thanks!

**Reply by bookvik** [reply]
![image](https://user-images.githubusercontent.com/204606/113279829-3b507d80-92ec-11eb-8721-65cd4dc5db85.png)

however, as it turned out, the tasklet jdbs does not have a prefix io it is inside the system

**Reply by bookvik** [reply]
I can help you make it/ exp: create wiki for this case

**Reply by mikub** [reply]
> ![image](https://user-images.githubusercontent.com/204606/113279829-3b507d80-92ec-11eb-8721-65cd4dc5db85.png)
> 
> however, as it turned out, the tasklet jdbs does not have a prefix io it is inside the system

Yep, good catch, the "io" is  in most of the other namespaces for other tasklets, but not in jdbc tasklet, my bad :)

**Comment by mikub** [comment]
@bookvik also if you want to use that Postgres DB as an **archive** storage for archived workflow jobs, see how it is configured e.g. here:
https://github.com/mikub/titanoboa/wiki/Server-Configuration#server-properties
(see the more complex example with archival system)

As a part of system config you should include `:archive-ds-ks` that points again to data source object within given DB connection pool,
usually it will be [\<archival system key\>  :system :db-pool] - so in that particular example it is `[:archival-system :system :db-pool]`
